File: grpc-pg-bench/spring-vt/src/main/java/com/beam/bench/Fnv.java
package com.beam.bench;

/**
 * 32-bit FNV-1a checksum, byte-identical to Go's {@code hash/fnv.New32a()} and
 * the other stacks. Java {@code int} arithmetic wraps mod 2^32, matching uint32
 * semantics, so every stack computes the same checksum for the same bytes.
 */
final class Fnv {
    private static final int OFFSET_BASIS = 0x811c9dc5; // 2166136261
    private static final int PRIME = 0x01000193;        // 16777619

    private Fnv() {
    }

    static int fnv1a32(byte[] bytes) {
        int hash = OFFSET_BASIS;
        for (byte b : bytes) {
            hash ^= (b & 0xff);
            hash *= PRIME;
        }
        return hash;
    }
}

File: grpc-pg-bench/spring-kt-vt/src/main/kotlin/com/beam/bench/Fnv.kt
package com.beam.bench

/**
 * 32-bit FNV-1a checksum, byte-identical to Go's `hash/fnv.New32a()` and every
 * other stack. Int arithmetic wraps mod 2^32, matching uint32 semantics, so all
 * stacks compute the same checksum for the same bytes.
 *
 * Reference vectors: fnv1a32("") = 2166136261, fnv1a32("a") = 0xe40c292c,
 * fnv1a32("foobar") = 0xbf9cf968.
 */
internal fun fnv1a32(bytes: ByteArray): Int {
    var hash = -0x7ee3623b // 2166136261 (FNV offset basis) as signed Int
    for (b in bytes) {
        hash = hash xor (b.toInt() and 0xff)
        hash *= 0x01000193 // FNV prime (16777619), Int multiply wraps mod 2^32
    }
    return hash
}
